### ISSUE TITLE
GLM Methods Section

### ISSUE THREAD

**Issue by alexpiet** [AllenInstitute/visual_behavior_glm#361]
Using this to draft a methods section

**Kernel Regression** 
Our regression model is linear kernel regression with a gaussian noise model. Each model feature is a 1 dimensional vector f_i(t), and was convolved with a learned kernel k_i(t) to produce the predicted model component of that feature f_i(t)*k_i(t). Model components were summed together to produce the full model response. r(t) = \sum_i f_i(t)*k_i(t). The learned kernels were instantiated as vectors of weights. The length of each feature kernel was determined using domain knowledge and trial and error. 

**Stimulus pre-processing** 
For each cell, the model was fit to either df/f traces, or L0 detected events smoothed with a causal half-gaussian filter (width?). To remove aliasing between the neural activity timestamps and feature timestamps all continuous signals were interpolated onto stimulus-aligned timestamps. Discrete features were binned onto the nearest stimulus-aligned timestamp. For a given experiment, if a feature had less than 5 discrete events, that feature was not included in the model. Continuous features like pupil diameter and running speed were standardized to have unit variance. 

**Toeplitz Matrix Implementation**
Convolutions are time-invariant linear shift operators, which allowed us to implement the convolution of kernels with feature timeseries as a banded toeplitz matrix operation. y=Wx. Where y is the 1d time-series for a cell, x is a vector of all kernel weights concatenated together, and W is a toeplitz matrix with diagonal bands that map kernel weights onto time-shifted features. (More details here). 

**Closed form solution and ridge regularization**
The Toeplitz matrix implementation with a gaussian noise model yields the standard solution for ordinary least squares regression: (equation here). We added a L2 ridge regression penalty to the cost function, resulting in the following solution: (equation here), where (lambda) is the L2 penalty. We fit the model using 5-fold cross validation. We split each session into 50 intervals, and randomly assigned 10 intervals to each cross validation fold. Therefore each of the 5 folds were intermingled in time. To determine the hyper-parameter (lambda) we evaluated the model on a grid of potential lambda values from 0 to 500, and for each cell selected the lambda that resulted in the best test-set performance. The training/test splits were different for hyper-parameter selection than fitting the model for analysis. 

**Model evaluation**
We evaluate our model by computed the explained variance: (var-total - var-residual)/var-total. If the smoothed L0 event trace had no activity (no events) on a cross-validation fold, then then cell would have no variance. Therefore, for numerical stability, when a cell had no activity on a cross-validation fold, the model explained variance was set to 0. 

To evaluate the contribution of each kernel in explaining each cell's activity we fit a series of reduced model where individual kernels or groups of kernels were removed from the model and the model was refit. We compute a dropout score which measures the fraction of the full model's explained variance that is lost when using the reduced model. This dropout score therefore captures the unique explained variance from that kernel or set of kernels. The raw dropout score is: -(1-VE_reduced_model)/VE_full_model. Some model features are only present during a portion of the session, we therefore adjust the raw dropout score to only evaluate the explained variance on the time points when that feature was present. The adjusted dropout score for feature (i) is: -(1-VE_reduced_model(t_i))/VE_full_model(t_i). Where t_i indicates that feature i was included on that timepoint (need better notation). For numerical stability, the dropout scores were bounded between (-1,0). Further, if the full model explained less than 0.5% of the variance on the relevant time points, the dropout score was set to 0. We therefore only analyzed cells with a full model explained variance above 0.5%
